source/distributed/fully_distributed_tria_util.cc

```diff
@@ -512,6 +514,97 @@ namespace parallel
 
         return construction_data;
       }
+
+
+
+      template <int dim, int spacedim>
+      ConstructionData<dim, spacedim>
+      create_construction_data_from_triangulation_in_groups(
+        std::function<void(dealii::Triangulation<dim, spacedim> &)>
+                                                serial_grid_generator,
+        std::function<void(dealii::Triangulation<dim, spacedim> &,
+                           const MPI_Comm,
+                           const unsigned int)> serial_grid_partitioner,
+        const MPI_Comm                          comm,
+        const int                               group_size,
+        const bool                              construct_multilevel_hierarchy)
+      {
+        const unsigned int my_rank =
+          dealii::Utilities::MPI::this_mpi_process(comm);
+        const unsigned int group_root = (my_rank / group_size) * group_size;
+
+        // check if process is root of the group
+        if (my_rank == group_root)
+          {
+            // Step 1: create serial triangulation
+            dealii::Triangulation<dim, spacedim> tria(
+              construct_multilevel_hierarchy ?
+                dealii::Triangulation<dim, spacedim>::none :
+                dealii::Triangulation<dim, spacedim>::
+                  limit_level_difference_at_vertices);
+            serial_grid_generator(tria);
+
+            // Step 2: partition active cells and ...
+            serial_grid_partitioner(tria, comm, group_size);
+
+            // ... cells on the levels if multigrid is required
+            if (construct_multilevel_hierarchy)
+              GridTools::partition_multigrid_levels(tria);
+
+
+            const unsigned int max_rank_in_group =
```

Comment:
```suggestion
            const unsigned int end_group =
```
You use this as the upper end of a half-open interval, whereas `max_rank_in_group` would suggest that it is still included.